compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/MainMethodGenerationLowering.kt

```diff
@@ -0,0 +1,97 @@
+import org.jetbrains.kotlin.ir.declarations.IrDeclarationOrigin
+import org.jetbrains.kotlin.ir.declarations.IrSimpleFunction
+import org.jetbrains.kotlin.ir.expressions.impl.IrConstImpl
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.util.getAnnotation
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.resolve.DescriptorUtils
+import org.jetbrains.kotlin.types.Variance
+
+internal class MainMethodGenerationLowering(val context: JvmBackendContext) : ClassLoweringPass {
+
+    override fun lower(irClass: IrClass) {
+        if (!context.configuration.languageVersionSettings.supportsFeature(LanguageFeature.ExtendedMainConvention)) return
+        if (irClass.origin != IrDeclarationOrigin.FILE_CLASS) return
+
+        val parameterlessMain = irClass.functions.find { it.isParameterlessMainMethod() }
+        val hasRegularMain = irClass.functions.any { it.isMainMethod() }
+
+        if (parameterlessMain == null || hasRegularMain) return
+
+        generateMainMethod(irClass, parameterlessMain)
+    }
+
+    private fun generateMainMethod(irClass: IrClass, parameterlessMain: IrSimpleFunction) {
+        irClass.addFunction {
+            name = Name.identifier("main")
+            visibility = Visibilities.PUBLIC
+            returnType = context.irBuiltIns.unitType
+            modality = Modality.FINAL
+            origin = JvmLoweredDeclarationOrigin.GENERATED_MAIN_FOR_PARAMETERLESS_MAIN_METHOD
+        }.apply {
+            addValueParameter {
+                name = Name.identifier("args")
+                type = context.irBuiltIns.arrayClass.typeWith(context.irBuiltIns.stringType)
+            }
+            body = context.createIrBuilder(this.symbol).irBlockBody {
+                +irReturn(this.irCall(parameterlessMain))
+            }
+        }
+    }
+
+    private fun IrSimpleFunction.isParameterlessMainMethod(): Boolean {
+        if (typeParameters.isNotEmpty()) return false
+        if (valueParameters.size != 0) return false
+        if (!returnType.isUnit()) return false
+        if (name.asString() != "main") return false
+        return true
+    }
+
+    private fun IrSimpleFunction.isMainMethod(): Boolean {
+        if (!hasJvmNameMain) return false
+        if (!returnType.isUnit()) return false
+        if (allParameters.size != 1) return false
+
+        val parameter = allParameters.first()
+        if (!parameter.type.isArray()) return false
+
+        val argType = (parameter.type as IrSimpleType).arguments.first()
+        return when (argType) {
+            is IrTypeProjection -> {
+                (argType.variance != Variance.IN_VARIANCE) && argType.isSubtypeOfNullableString
+            }
+            else -> false
+        }
+    }
+
+    private val IrTypeProjection.isSubtypeOfNullableString
+        get() = type.isSubtypeOf(context.irBuiltIns.stringType.makeNullable(), context.irBuiltIns)
```

Comment:
Probably would be more correct to just call `type.isStringClassType()`, and inline this function, to prevent `Array<Nothing>` from being recognized as a string array

Reply:
Done.